Enable parsing columns from file path for Broker Load (#1582)

yuanlihan opened the pull request:
Currently, we do not support parsing encoded/compressed columns in file path, eg: extract column k1 from file path `/path/to/dir/k1=1/xxx.csv`

This patch is able to parse columns from file path like in Spark(Partition Discovery).

This patch parse partition columns at BrokerScanNode.java and save parsing result of each file path as a property of TBrokerRangeDesc, then the broker reader of BE can read the value of specified partition column.

yuanlihan committed:
Enable parsing columns from file path for Broker Load

yuanlihan committed:
Enable parsing columns from file path for Broker Load

kangpinghuang committed:
Fix tablet restore api in BE(#1623) (#1624)

wangbo committed:
Insert select Stmt keep the same semantics with mysql (#1626) (#1628)

imay committed:
Fix compile error (#1630)

EmmyMiao87 committed:
Change the prefix of table info apis (#1625)

The pathtrie could not distinguish the different param key with the same prefix path.
So the prefix of table info apis has been change to /api/external which is used by spark-doris-connector.

imay commented on be/src/exec/broker_scanner.cpp:
```diff
@@ -489,18 +504,18 @@ bool BrokerScanner::line_to_src_tuple(const Slice& line) {
-        if (slot_desc->is_nullable() && is_null(value)) {
-            _src_tuple->set_null(slot_desc->null_indicator_offset());
-            continue;
+        auto iter = _columns_from_path.find(slot_desc->col_name());
```
Could you find this one time when this class is initialized?
If you did like that, we can reduce find operation here for each filed of each row.

imay commented on be/src/exec/broker_scanner.cpp:
```diff
@@ -489,18 +504,18 @@ bool BrokerScanner::line_to_src_tuple(const Slice& line) {
+        auto iter = _columns_from_path.find(slot_desc->col_name());
+        if (iter != _columns_from_path.end()) {
+            std::string partitioned_field = iter->second;
+            const Slice value = Slice(partitioned_field.c_str(), partitioned_field.size());
```
What you reference `partitioned_field` is a temporary variable. It is not valid out of code block.

imay commented on be/src/exec/parquet_reader.cpp:
```diff
@@ -204,13 +208,23 @@ Status ParquetReaderWrap::read(Tuple* tuple, const std::vector<SlotDescriptor*>&
             auto slot_desc = tuple_slot_descs[i];
-            column_index = i;// column index in batch record
-            switch (_parquet_column_type[i]) {
+            auto iter = _columns_from_path.find(slot_desc->col_name());
```
I think we can finish it in BrokerScanner's line_to_src_tuple. Which can avoid same code in different places.

yuanlihan replied:
Currently, we have two broker scanners. One is BrokerScanner using for reading CSV like row-storage sources. The other is ParquetScanner using for read parquet files, which is a column-storage source. It seems that we cannot reuse the `line_to_src_tuple` method of BrokerScanner

imay replied:
I think we can write a common function to fill partition columns which can be used in both two scanners, like `fill_dest_tuple`. If this function is not easy to write and used by scanners, it's OK to write one copy for each.

imay commented on fe/src/main/java/org/apache/doris/planner/BrokerScanNode.java:
```diff
@@ -696,9 +699,40 @@ private TBrokerRangeDesc createBrokerRangeDesc(long curFileOffset, TBrokerFileSt
+        Map<String, String> columns = new HashMap<>();
+        for (int i = strings.length - 1; i >= 0; i--) {
+            String str = strings[i];
+            if (str == null || str.isEmpty() || !str.contains("=")) {
```
When we meet a path like "/xx/yy/zz/k1=1/aa",  I think we should think this path is not a valid path. We only accept paths like "xx/yy/zz/aa/k1=1"

yuanlihan replied:
> When we meet a path like "/xx/yy/zz/k1=1/aa", I think we should think this path is not a valid path. We only accept paths like "xx/yy/zz/aa/k1=1"

I will raise an exception here.

imay commented on fe/src/main/java/org/apache/doris/planner/BrokerScanNode.java:
```diff
@@ -696,9 +699,40 @@ private TBrokerRangeDesc createBrokerRangeDesc(long curFileOffset, TBrokerFileSt
+        if (columns.size() != columnsFromPath.size()) {
+            throw new UserException("Fail to parse columnsFromPath, expected: " + columnsFromPath + ", filePath: " + filePath);
+        }
+        LOG.info("columns from path: " + columns + ", filePath: " + filePath);
```
You should remove this log

imay committed:
Fix HLLContext cast error (#1632)

yuanlihan commented on fe/src/main/cup/sql_parser.cup:
```diff
@@ -1050,10 +1050,11 @@ data_desc ::=
     opt_partitions:partitionNames
     opt_field_term:colSep
     opt_file_format:fileFormat
+    opt_columns_from_path:columnsFromPath
```
> Could you please put this behind colList?

Sure

Commit:
Change cumulative compaction for decoupling storage from compution (#1576)

1. Calculate cumulative point when loading tablet first time.
2. Simplify pick rowsets logic upon delete predicate.
3. Saving meta and modify rowsets only once after cumulative compaction.

HangyuanLiu committed:
Fix time zone compatibility (#1631)

yuanlihan committed:
Enable parsing columns from file path for Broker Load

yuanlihan committed:
Merge branch 'parse_columns_from_file_path' of https://github.com/yuanlihan/incubator-doris into parse_columns_from_file_path

yuanlihan committed:
Enable parsing columns from file path for Broker Load

yuanlihan committed:
Enable parsing columns from file path for Broker Load

yuanlihan committed:
Enable parsing columns from file path for Broker Load
